cmd/ingester/app/consumer/consumer.go

```diff
@@ -29,153 +28,116 @@ import (
+	msgMetrics := h.consumer.newMsgMetrics(claim.Partition())
 
-type consumerState struct {
-	wg                sync.WaitGroup
-	partitionConsumer sc.PartitionConsumer
+	for {
+		select {
+		case msg, ok := <-claim.Messages():
+			if !ok {
+				h.logger.Info("Message channel closed. ", zap.Int32("partition", claim.Partition()))
+				return nil
+			}
+			h.logger.Debug("Got msg", zap.Any("msg", msg))
+			msgMetrics.counter.Inc(1)
+			msgMetrics.offsetGauge.Update(msg.Offset)
+			msgMetrics.lagGauge.Update(claim.HighWaterMarkOffset() - msg.Offset - 1)
+			msgProcessor := h.getProcessFactory(sess, claim.Partition(), msg.Offset)
+			msgProcessor.Process(&saramaMessageWrapper{msg})
+		}
+	}
+	return nil
 }
 
 // New is a constructor for a Consumer
 func New(params Params) (*Consumer, error) {
-	deadlockDetector := newDeadlockDetector(params.MetricsFactory, params.Logger, params.DeadlockCheckInterval)
 	return &Consumer{
 		metricsFactory:      params.MetricsFactory,
 		logger:              params.Logger,
 		internalConsumer:    params.InternalConsumer,
 		processorFactory:    params.ProcessorFactory,
-		deadlockDetector:    deadlockDetector,
-		partitionIDToState:  make(map[int32]*consumerState),
 		partitionsHeldGauge: partitionsHeldGauge(params.MetricsFactory),
 	}, nil
 }
 
 // Start begins consuming messages in a go routine
 func (c *Consumer) Start() {
-	c.deadlockDetector.start()
 	go func() {
 		c.logger.Info("Starting main loop")
-		for pc := range c.internalConsumer.Partitions() {
-			c.partitionMapLock.Lock()
-			if p, ok := c.partitionIDToState[pc.Partition()]; ok {
-				// This is a guard against simultaneously draining messages
-				// from the last time the partition was assigned and
-				// processing new messages for the same partition, which may lead
-				// to the cleanup process not completing
-				p.wg.Wait()
+		ctx := context.Background()
+		handler := consumerGroupHandler{
+			processorFactory: c.processorFactory,
+			logger:           c.logger,
+			consumer:         c,
+		}
+		defer func() { _ = c.internalConsumer.Close() }()
+
+		go func() {
+			for err := range c.internalConsumer.Errors() {
+				if error, ok := err.(*sarama.ConsumerError); ok {
+					c.logger.Info("Starting error handler", zap.Int32("partition", error.Partition))
+					errMetrics := c.newErrMetrics(error.Partition)
+					errMetrics.errCounter.Inc(1)
+					c.logger.Error("Error consuming from Kafka", zap.Error(err))
+				}
 			}
-			c.partitionIDToState[pc.Partition()] = &consumerState{partitionConsumer: pc}
-			c.partitionIDToState[pc.Partition()].wg.Add(2)
-			c.partitionMapLock.Unlock()
-			c.partitionMetrics(pc.Partition()).startCounter.Inc(1)
-			go c.handleMessages(pc)
-			go c.handleErrors(pc.Partition(), pc.Errors())
+			c.logger.Info("Finished handling errors")
+		}()
+
+		for {
+			err := c.internalConsumer.Consume(ctx, &handler)
+			if err != nil {
+				panic(err)
```

Comment:
That's rude. Do you really want a panic here?

Reply:
Ha, not really.